rpm/files.go

```diff
@@ -0,0 +1,118 @@
+package rpm
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"sync"
+
+	"github.com/quay/claircore"
+	"github.com/quay/zlog"
+)
+
+// filesCache is used for concurrent access to the map containing
+// [claircore.Layer] -> map[string]struct{}.
+type filesCache struct {
+	c    map[string]map[string]struct{}
+	refs map[string]int
+	mu   sync.Mutex
+}
+
+var fc = &filesCache{
+	c:    map[string]map[string]struct{}{},
+	refs: map[string]int{},
+}
+
+// gc decrements the reference counter and deletes the [claircore.Layer]'s
+// entry from the cache map and the reference counter once the references
+// are 0 (or less).
+func (fc *filesCache) gc(key string) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.refs[key]--
+	if fc.refs[key] <= 0 {
+		delete(fc.c, key)
+		delete(fc.refs, key)
+	}
+}
+
+// get increments the reference counter for the key and looks up the key
+// in the cache. It will subsequently start a goroutine to track when the
+// calling context is done, calling the gc method.
+func (fc *filesCache) get(ctx context.Context, key string) (map[string]struct{}, bool) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.refs[key]++
+	f, ok := fc.c[key]
+	go func() {
+		<-ctx.Done()
```

Comment:
Can you confirm the contexts will actually close and they aren't just `context.Background()`?

Reply:
https://github.com/quay/claircore/blob/main/libindex/libindex.go#L187 The context that gets passed to the indexing-related components should be `Done` after the indexing process finishes.